### ISSUE TITLE
WRC 4 [PAL] - black screen past PlayStation 2 logo

### ISSUE THREAD

**Issue by ghost** [PCSX2/pcsx2#1782]
***PCSX2 version:***
v1.5.0-dev-1789

***PCSX2 options:***
Safest with no MTVU.

***Plugins used:***
Default plugins. GSdx32-SSE4.

***Description of the issue:***
(Both full boot and fast boot do the same thing-ish. The below descriptions were replicated using full boot.)

(DX11 Hardware) Once booting past the PlayStation 2 logo, the game shows a black screen at 100% speed. After about 20 seconds the game will lock up at a slow speed and cause PCSX2 to lock up and not respond. https://cloud.githubusercontent.com/assets/10300935/22042191/fdb326a8-dd6e-11e6-8ba7-118cf0795711.PNG) & (https://cloud.githubusercontent.com/assets/10300935/22041969/e7313e5c-dd6d-11e6-8cfe-f9b7611e4a77.PNG)
 Loads of TLB errors flood through the console afterwards.

(DX11 software) The game will do nothing with a frozen PlayStation 2 logo halfway through the bottom of the window continuously at 100% speed. (https://cloud.githubusercontent.com/assets/10300935/22042080/80b9f5be-dd6e-11e6-8305-b5a632793a09.PNG) & (https://cloud.githubusercontent.com/assets/10300935/22042066/6fd2e6a2-dd6e-11e6-9c3a-3754c621dd23.PNG) Loads more TLB errors flood through the console afterwards as well, of course.

***How to reproduce the issue:***
Full boot the WRC 4 ISO (hacks don't make any difference to the behavior as far as I know).

***Last known version to work:***
According to the wiki, the game won't load due to TLB miss errors. The last known version to work was PCSX2 0.9.4. (http://wiki.pcsx2.net/index.php/WRC_4)

***PC specifications:***
CPU: Intel i5 2320
GPU: ASUS STRIX GeForce GTX 1060 6GB
OS: Windows 10 1607 64-bit

**Comment by refractionpcsx2** [comment]
From what I've seen/heard, this broke around the time the memory management moved away from the VM/TLB setup we used to have, I think then only the TLB version worked.

**Comment by gregory38** [comment]
Could you test the EE interpreter ?

**Comment by refractionpcsx2** [comment]
I believe the interpreter doesn't work either now.

**Comment by FlatOutPS2** [comment]
No setting works. This is a really old issue. From what I read it seems to require RAM caching.

**Comment by gregory38** [comment]
The interpreter handles the TLB correctly. I wanted to be sure it wasn't a basic TLB issue. Hum RAM caching, could also be the branch in delay slot and such nicety. Hopefully your PR will handle the delay slot.

**Comment by atomic83GitHub** [comment]
I have the demo of the game and I had the same issue , but the interpretor cause a black screen at the lunch of this game (but no crash) and TLB error are gone but the result is the same :( .

**Comment by atomic83GitHub** [comment]
And by the way I will report later that the SCEE Hits Demo have a lot of TLB issue.

**Comment by kozarovv** [comment]
Hello, i took a look at this game. Something tricky going on here with memory management of pcsx2. 
Game seems to use memory range above 0x20000000, and on real PS2 is fine somehow. 
Issue here is that unpacker deletes itself while trying to write to 0x204xxxxx, but that write also to 0x4xxxxx which probably won't happen on real PS2. No idea what even is 0x20000000, and why Ps2 game have access to it. 

Here is why WRC3 work: 
Game load (i nored values for easier read): 
![assert (1)](https://user-images.githubusercontent.com/15552250/82764778-afd7a780-9e11-11ea-9253-6a7a11a095fb.jpg)
When unpack is done, at 0x20100000 + 0x2273C0 = 0x203273C0 (stored now in t4). Condition to out of loop is met, and game load from 0x100008
![unknown](https://user-images.githubusercontent.com/15552250/82764763-8e76bb80-9e11-11ea-811d-265217e5c41c.png)

Here is why WRC4 fail: 
![gsdx0 94](https://user-images.githubusercontent.com/15552250/82764867-7a7f8980-9e12-11ea-93d6-6e4f8e3ab29b.jpg)
Like you can see, end of loop should happen on 0x204A3340. This condition can't be made because on pcsx2 writes above 0x20000000 overwrite real memory at 0x00. So write to 0x204A3340 never gonna happen as before that 0x400XXX get overwritten. This make unpacker to overwrite itself, and few cycles later overall mess. 

Not sure that i'm 100% understandable here. Fact that WRC series work is pure luck, because unpacked code is always below 0x400000, WRC4 is just greater part of code, and shows some memory mirroring issue in pcsx2. Not sure that is even "proper" issue from emulation side, but that's what happen.

**Comment by PSI-Rockin** [comment]
I know what the issue is. 0x2xxxxxxx is an uncached region of RAM - there's nothing wrong with writing to it. Even on real hardware, the decompressor overwrites its memory. However, the game works because the decompressor is stored in the icache, and all writes are being done to main RAM only. The icache is large enough to fit the whole decompressor, so nothing bad happens. PCSX2 does not handle the icache, and so the decompressor gets overwritten with junk.

tl;dr for this game to work without a hack, icache emulation is needed.

**Comment by kozarovv** [comment]
Ah, ok. So i think that i can post here this hack for people that want to just play it. This is far away from proper fix, and from good solution. But i tested it, and is working fine to load game fine. 

This fix work for WRC 4 PAL, but only for NON PLATINIUM release (md5: 4dea5a6f8509748e12e1982f4d6a6df0).

This is ppf patcher + patch in zip. To use, simple select ISO file, and included ppf patch file, click APPLY button. Done. 
[fix.zip](https://github.com/PCSX2/pcsx2/files/4675360/fix.zip)

I moved unpacker, and entry point to 0x600000. Fixed 1 jump, and that's all. Game boot now. 
Remember to enable VU XGKICK hack. 

![assert](https://user-images.githubusercontent.com/15552250/82781845-2b117b80-9e5b-11ea-8418-6b4d3432dcde.jpg)

**Comment by kozarovv** [comment]
I created pnach to make game boot **without need to patch ISO**. I think there is nothing better that can be done without emulation of iCache. Patch can be downloaded here: https://forums.pcsx2.net/Thread-Fixing-unplayable-games?pid=609793#pid609793

**Comment by Tasosgemah** [comment]
> I created pnach to make game boot **without need to patch ISO**. I think there is nothing better that can be done without emulation of iCache. Patch can be downloaded here: https://forums.pcsx2.net/Thread-Fixing-unplayable-games?pid=609793#pid609793

I know this is an old comment but how do you use thie patch file? I only see some text in the link.

**Comment by mirh** [comment]
https://forums.pcsx2.net/Thread-How-PNACH-files-work-2-0

**Comment by stenzek** [comment]
You don't need to use any patch files. The patch for the unpacker is included in the Game Database.

**Comment by mirh** [comment]
I see, fixed by #3829 and #9085
